### PR TITLE
Replace built-in docs codeblock's leading spaces with tabs

### PR DESCRIPTION
**Pull request by Mickeon**
This PR replaces the spaces used in the codeblocks of the class references with actual tabs.
If your Editor Setting's preference is using spaces, nothing changes.
With one GIF, this means the following (notice the tabs):

| Before | After |
| --- | --- |
|  | 

However, hm... It comes at a cost. For the tab to display in a way that actually looks good, the **RichTextLabel**'s `tab_size` has to be set to `8`, otherwise the tab's indent would be way, way too small, for some reason. Because of this, there's the side effect of increasing the indentation across the whole built-in documentation:

Take a closer look:

| Before | After |
| --- | --- |
| image | image |
| image |image | 
| image | image |
| image | image

Some may say this is a happy accident, some may absolutely not. So keep that in mind.

### CODE REVIEW

**Comment by YuriSizov** [comment]
Could you please rebase on the current master? We had an issue with macOS builds which failed your CI.

**Comment by Mickeon** [comment]
Done just that.

**Comment by MewPurPur** [comment]
I don't like the increased indent but I think it's worth it. I'm just confused why we'd need an 8 tab size for what's supposed to be monospace. I would've expected the mono tabs to be as wide as my code editor theme (which is at 3 for me).

**Comment by Mickeon** [comment]
> I'm just confused why we'd need an 8 tab size for what's supposed to be monospace.

Honestly, I am not sure why **RichTextLabel** acts like this. I tried a value of `4` which you would think would be appropriate, but it resulted in the padding being **way too** short for some reason.

It originally was also based on the "indent size" Editor Setting, but because it would've affected the entire **RichTextLabel** I was not sure about it.

File: editor/editor_help.cpp [hunk 1]
```diff
@@ -2061,6 +2062,24 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, Control
 
 		if (brk_pos > pos) {
 			String text = bbcode.substr(pos, brk_pos - pos);
+			if (codeblock_tag && code_indent_type == 0) {
+				// Replace the code block's space indentation with tabulation.
+				PackedStringArray text_lines = text.split("\n");
+				text = String();
+				for (String line : text_lines) {
```
Comment (AThousandShips) [inline editor/editor_help.cpp:2069]:
```suggestion
				for (const String &line : text_lines) {
```

File: editor/editor_help.cpp [hunk 2]
```diff
@@ -2061,6 +2062,24 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, Control
 
 		if (brk_pos > pos) {
 			String text = bbcode.substr(pos, brk_pos - pos);
+			if (codeblock_tag && code_indent_type == 0) {
+				// Replace the code block's space indentation with tabulation.
```
Comment (AThousandShips) [inline editor/editor_help.cpp:2066]:
```suggestion
				// Replace the code block's space indentation with tabs.
```
"tabulation" means creating a list

**Review by AThousandShips — APPROVED** [review]
LGTM

**Review by Calinou — DISMISSED** [review]
Tested locally, there are issues with code samples that contain multiple tabs on the same line.

For instance, the HTTPRequest class description has a code sample where the `push_error()` line has two leading tabs, but none are displayed and copied here:

image

If you select the block and paste it in a script, you get this:

image

**Comment by Mickeon** [comment]
You will never guess who the idiot that swapped divisor and dividend around is.

**Comment by Mickeon** [comment]
You will never guess who the idiot that completely f<sub>u</sub><sup>c</sup><sub>k</sub><sup>e</sup>d up the formula is.

Should work fine now:

File: editor/editor_help.cpp [hunk 3]
```diff
@@ -2384,6 +2384,7 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, Control
 	List<String> tag_stack;
 	bool code_tag = false;
 	bool codeblock_tag = false;
+	const int code_indent_type = EDITOR_GET("text_editor/behavior/indent/type");
```
Comment (KoBeWi) [inline editor/editor_help.cpp:2387]: Changing this setting won't update currently opened help pages.
Not sure if it's worth handling though.

File: editor/editor_help.cpp [hunk 4]
```diff
@@ -2395,6 +2396,24 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, Control
 
 		if (brk_pos > pos) {
 			String text = bbcode.substr(pos, brk_pos - pos);
+			if (codeblock_tag && code_indent_type == 0) {
```
Comment (KoBeWi) [inline editor/editor_help.cpp:2399]: Could use a comment on where the 0 comes from.

File: editor/editor_help.cpp [hunk 5]
```diff
@@ -2395,6 +2396,24 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, Control
...
+					String stripped_line = line.dedent();
+					int space_count = line.length() - stripped_line.length();
+
+					if (!text.is_empty()) {
+						text += "\n";
+					}
+					if (space_count > 0) {
+						text += String("\t").repeat(MAX(space_count / 4, 1)) + stripped_line;
+					} else {
+						text += line;
```
Comment (KoBeWi) [inline editor/editor_help.cpp:2413]: Looks like a job for StringBuilder.

File: editor/editor_help.cpp (hunk 4 above)
Comment (Mickeon) [reply editor/editor_help.cpp:2399]: I'd rather modify this part here to use a bool.

In a prior version of this PR I also handled spaces and adjusted the padding depending on `text_editor/behavior/indent/size`, but in practice it looked very bad.

**Review by KoBeWi — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by Mickeon** [comment]
And from here on out begins the built-in reference revolution

**Pull request merged**